src/drivers/heater/heater.cpp

```diff
@@ -0,0 +1,345 @@
+	px4_arch_configgpio(GPIO_HEATER_OUTPUT);
+	px4_arch_gpiowrite(GPIO_HEATER_OUTPUT, 0);
+}
+
+Heater::~Heater()
+{
+	_task_should_exit = true;
+	work_cancel(LPWORK, &_work);
+	px4_arch_gpiowrite(GPIO_HEATER_OUTPUT, 0);
+
+	// Check if GPIO is stuck on, and if so, configure it as an input pulldown then reconfigure as an output.
+	if (px4_arch_gpioread(GPIO_HEATER_OUTPUT)) {
+		px4_arch_configgpio(GPIO_HEATER_INPUT);
+		px4_arch_configgpio(GPIO_HEATER_OUTPUT);
+		px4_arch_gpiowrite(GPIO_HEATER_OUTPUT, 0);
+	}
+}
+
+void Heater::check_params(const bool force)
+{
+	bool updated;
+	parameter_update_s paramUpdate;
+
+	orb_check(_parameter_sub, &updated);
+
+	if (updated) {
+		orb_copy(ORB_ID(parameter_update), _parameter_sub, &paramUpdate);
+	}
+
+	if (updated || force) {
+		param_get(_p_target_temp, &_target_temp);
+	}
+}
+
+void
+Heater::heater_controller_trampoline(void *arg)
+{
+	Heater *dev = reinterpret_cast<Heater *>(arg);
+
+	dev->heater_controller();
+}
+
+void Heater::heater_controller()
+{
+	if (!_task_is_running) {
+		// Initialize uORB topics.
+		initialize_topics();
+
+		check_params(true);
+
+		// Task is now running, keep doing so until we need to stop.
+		_task_is_running = true;
+	}
+
+	if (_heater_on) {
+		// Turn the heater off.
+		px4_arch_gpiowrite(GPIO_HEATER_OUTPUT, 0);
+		_heater_on = false;
+
+		// Check if GPIO is stuck on, and if so, configure it as an input pulldown then reconfigure as an output.
+		if (px4_arch_gpioread(GPIO_HEATER_OUTPUT)) {
+			px4_arch_configgpio(GPIO_HEATER_INPUT);
+			px4_arch_configgpio(GPIO_HEATER_OUTPUT);
+			px4_arch_gpiowrite(GPIO_HEATER_OUTPUT, 0);
+		}
+	}
+
+	else {
+		check_params(false);
+		update_topics();
+
+		// Obtain the current IMU sensor temperature.
+		_current_temp = _sensor_accel.temperature;
+
+		// Calculate the error between controller setpoint and the current temperature.
+		_error_temp = _target_temp - _current_temp;
+
+		// Modulate the heater time on with a feedforward/PI controller.
+		_proportional_value = _error_temp * _proportional_gain;
+		_integrator_value += _error_temp * _integrator_gain;
+		_integrator_value = math::max(math::min(_integrator_value, 0.25f), -0.25f);
+
+		_controller_time_on_usec = (int)((_feed_forward + _proportional_value +
+						  _integrator_value) * (float)_controller_period_usec);
+
+		// Ensure the heater time on is clamped within the time allowed.
+		_controller_time_on_usec = math::max(math::min(_controller_period_usec, _controller_time_on_usec), 0);
+
+		// Employ a complimentary filter to smooth the duty cycle value over roughly a 2 second time constant.
+		_duty_cycle = (0.05f * ((float)_controller_time_on_usec / (float)_controller_period_usec)) + (0.95f * _duty_cycle);
+
+		// Turn the heater on.
+		_heater_on = true;
+		px4_arch_gpiowrite(GPIO_HEATER_OUTPUT, 1);
+	}
+
+
+	if (!_task_should_exit) {
+
+		// Schedule the next cycle.
+		if (_heater_on) {
+			work_queue(LPWORK, &_work, (worker_t)&Heater::heater_controller_trampoline, this,
```

Comment:
How time-sensitive is this? If we run on `LPWORK` you must expect that there can be times where the driver is not scheduled for some hundreds of ms. If you need more scheduling accuracy, we should use `HPWORK`.